flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.executiongraph.JobInformation;
+import org.apache.flink.runtime.executiongraph.NoOpExecutionDeploymentListener;
+import org.apache.flink.runtime.executiongraph.failover.flip1.partitionrelease.PartitionReleaseStrategyFactoryLoader;
+import org.apache.flink.runtime.io.network.partition.NoOpJobMasterPartitionTracker;
+import org.apache.flink.runtime.jobgraph.ScheduleMode;
+import org.apache.flink.runtime.jobgraph.tasks.CheckpointCoordinatorConfiguration;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.runtime.shuffle.NettyShuffleMaster;
+import org.apache.flink.runtime.state.SharedStateRegistry;
+import org.apache.flink.runtime.state.memory.MemoryStateBackend;
+import org.apache.flink.runtime.testingUtils.TestingUtils;
+import org.apache.flink.util.SerializedValue;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.runtime.checkpoint.CheckpointCoordinatorTestingUtils.mockExecutionVertex;
+import static org.apache.flink.runtime.scheduler.declarative.WaitingForResourcesTest.assertNonNull;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/** Tests for the {@link StopWithSavepoint} state. */
+public class StopWithSavepointTest extends TestLogger {
+
+    /** Test failure during savepoint creation: go to failing */
+    @Test
+    public void testFailureDuringSavepointCreationWithNoRestart() throws Exception {
+        MockExecutingStateWithFailureHandlerContext ctx =
+                new MockExecutingStateWithFailureHandlerContext();
+
+        StopWithSavepointEnvironment stopWithSavepointEnv = createStopWithSavepointEnvironment(ctx);
+        StopWithSavepoint stopWithSavepoint = stopWithSavepointEnv.getState();
+        stopWithSavepoint.onEnter();
+
+        ctx.setHowToHandleFailure(ExecutingStateWithFailureHandler.FailureResult::canNotRestart);
+        // fail future returned by the CheckpointCoordinator
+        stopWithSavepointEnv
+                .getSavepointFuture()
+                .completeExceptionally(new RuntimeException("Savepoint creation failed"));
+
+        stopWithSavepointEnv.getExecutionGraphTerminationFuture().complete(JobStatus.FAILED);
+
+        ctx.setExpectFailing(assertNonNull());
+
+        ctx.close(); // trigger outstanding executions
+
+        assertThat(
+                stopWithSavepointEnv.getOperationCompletionFuture().isCompletedExceptionally(),
```

Comment:
I am not a huge fan of this environment. Looking alone on this test, it is not clear to me why the env has the operation completion future. If the operation future belonged to the state, then it would be clearer, for example.